File: engine/src/main/java/com/drake/engine/component/BasePagerAdapter.java
/*
 * Copyright (C) 2018, Umbrella CompanyLimited All rights reserved.
 * Project：Engine
 * Author：Drake
 * Date：9/11/19 7:25 PM
 */

package com.drake.engine.component;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.viewpager.widget.PagerAdapter;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.drake.engine.base.EngineKt.App;

public class BasePagerAdapter extends PagerAdapter {

  private List<String> titles = new ArrayList<>();
  private List<ViewDataBinding> viewDataBindings = new ArrayList<>();

  /**
   * 布局
   */
  public BasePagerAdapter(Integer[] layoutIds) {
    for (Integer layoutId : layoutIds) {
      ViewDataBinding viewDataBinding = DataBindingUtil.inflate(LayoutInflater.from(App), layoutId, null, false);
      viewDataBindings.add(viewDataBinding);
    }
  }

  public BasePagerAdapter() {
  }

  /**
   * 标题和布局
   */
  public BasePagerAdapter(String[] titles, Integer[] layoutIds) {
    this.titles.addAll(Arrays.asList(titles));
    for (Integer layoutId : layoutIds) {
      ViewDataBinding viewDataBinding = DataBindingUtil.inflate(LayoutInflater.from(App), layoutId, null, false);
      viewDataBindings.add(viewDataBinding);
    }
  }


  @Override
  public int getCount() {
    return viewDataBindings.size();
  }

  @NonNull
  @Override
  public Object instantiateItem(@NonNull ViewGroup container, int position) {
    ViewDataBinding viewDataBinding = viewDataBindings.get(position);
    View view = viewDataBinding.getRoot();
    container.addView(view);
    return view;
  }

  @Override
  public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
    container.removeView((View) object);
  }

  @Override
  public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
    return view == object;
  }

  @Nullable
  @Override
  public CharSequence getPageTitle(int position) {
    if (titles != null) {
      return titles.get(position);
    } else {
      return null;
    }
  }

  public <B extends ViewDataBinding> B getViewDataBinding(int position) {
    return (B) viewDataBindings.get(position);
  }

  /**
   * 通过布局创建并且返回ViewDataBinding
   */
  public <B extends ViewDataBinding> B add(int layoutId) {
    ViewDataBinding viewDataBinding = DataBindingUtil.inflate(LayoutInflater.from(App), layoutId, null, false);
    viewDataBindings.add(viewDataBinding);
    return (B) viewDataBinding;
  }

  public <B extends ViewDataBinding> B add(int layoutId, String title) {
    ViewDataBinding viewDataBinding = DataBindingUtil.inflate(LayoutInflater.from(App), layoutId, null, false);
    viewDataBindings.add(viewDataBinding);
    titles.add(title);
    return (B) viewDataBinding;
  }
}

File: engine/src/main/java/com/drake/engine/callback/BasePagerAdapter.kt
/*
 * Copyright (C) 2018, Umbrella CompanyLimited All rights reserved.
 * Project：Engine
 * Author：Drake
 * Date：9/11/19 7:25 PM
 */

package com.drake.engine.callback

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.viewpager.widget.PagerAdapter
import com.drake.engine.base.App
import java.util.*

class BasePagerAdapter : PagerAdapter {

    private val titles = ArrayList<String>()
    private val viewDataBindings = ArrayList<ViewDataBinding>()

    /**
     * 布局
     */
    constructor(layoutIds: Array<Int>) {
        for (layoutId in layoutIds) {
            val viewDataBinding = DataBindingUtil.inflate<ViewDataBinding>(
                LayoutInflater.from(App),
                layoutId,
                null,
                false
            )
            viewDataBindings.add(viewDataBinding)
        }
    }

    /**
     * 标题和布局
     */
    constructor(titles: Array<String>, layoutIds: Array<Int>) {
        this.titles.addAll(Arrays.asList(*titles))
        for (layoutId in layoutIds) {
            val viewDataBinding = DataBindingUtil.inflate<ViewDataBinding>(
                LayoutInflater.from(App),
                layoutId,
                null,
                false
            )
            viewDataBindings.add(viewDataBinding)
        }
    }


    override fun getCount(): Int {
        return viewDataBindings.size
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val viewDataBinding = viewDataBindings[position]
        val view = viewDataBinding.root
        container.addView(view)
        return view
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as View)
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }

    override fun getPageTitle(position: Int): CharSequence? {
        return titles?.get(position)
    }

    fun <B : ViewDataBinding> getViewDataBinding(position: Int): B {
        return viewDataBindings[position] as B
    }

    /**
     * 通过布局创建并且返回ViewDataBinding
     */
    fun <B : ViewDataBinding> add(layoutId: Int): B {
        val viewDataBinding = DataBindingUtil.inflate<ViewDataBinding>(
            LayoutInflater.from(App),
            layoutId,
            null,
            false
        )
        viewDataBindings.add(viewDataBinding)
        return viewDataBinding as B
    }

    fun <B : ViewDataBinding> add(layoutId: Int, title: String): B {
        val viewDataBinding = DataBindingUtil.inflate<ViewDataBinding>(
            LayoutInflater.from(App),
            layoutId,
            null,
            false
        )
        viewDataBindings.add(viewDataBinding)
        titles.add(title)
        return viewDataBinding as B
    }
}
